### PR TITLE
documentation: Add github workflow guidelines

### PR DESCRIPTION
**Pull request by mathiaszinnen**

### CODE REVIEW

File: doc/github-workflow.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+- One for each repository, linked to issues
+- One project board for each student project in the corresponding repo, possibly linked to issues 
+  
+## Issues
+- One issue board for each repository
+- Issues should be tagged appropriately to enable external contributions
+
+## External Contributions
+- External contributions are to be encouraged
+- External contributors may stage issues or work on existing ones
```
Comment (georg-schwarz) [inline doc/github-workflow.md:15]: Self-staged issues may be discussed with the admin teams in order to ensure later integration.

Comment (georg-schwarz) [reply doc/github-workflow.md:15]: Irgendwie sowas noch?

File: doc/github-workflow.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+- For integration into ods codebase they need to sign a contributers agreement (via the CLA Assistant GitHub Plugin)
+- Code integration and review is to be done by pull requests. The pull requests have to be approved by at least one member of the repositories admins team. 
+
+## Repositories
+- Master branch is protected: No direct push, only via reviewed pull request from separate branches
+- If possible, pull requests should be linked to issues
+- Students work on their own branches using regular pull requests to get fast feedback
+- Separate student forks only if the student projects makes it necessary. This enables fast integration of changes and prevents extensive integration effort
```
Comment (georg-schwarz) [inline doc/github-workflow.md:23]: Not sure if it really makes a difference for us.... They could just fork and send PRs

Comment (mathiaszinnen) [reply doc/github-workflow.md:23]: In the end, this is probably a matter of taste. If you prefer the fork solution, I would suggest using that.

**Comment by andreas-bauer** [comment]
I'm very skeptical that this document brings a benefit to the project.

IMHO a `Contribution` and `Development` section in the README.md would be a better alternative.

Example;
```Markdown
## Contributing to the ODS
Contributions are welcome. Thank you if you wanna contribute to this project.
There are several ways how you can contribute to the ODS:
- by implementing new features 
- by fixing bugs
- by improving the documentation
- by discussing use cases that are not covered by the ODS yet

## Development
If a commit is related to an issue (e.g. closes an issue) it should contain the issue number in the commit.
```

**Comment by mathiaszinnen** [comment]
> I'm very skeptical that this document brings a benefit to the project.
> 
> IMHO a `Contribution` and `Development` section in the README.md would be a better alternative.
> 
> Example;
> 
> ```gfm
> ## Contributing to the ODS
> Contributions are welcome. Thank you if you wanna contribute to this project.
> There are several ways how you can contribute to the ODS:
> - by implementing new features 
> - by fixing bugs
> - by improving the documentation
> - by discussing use cases that are not covered by the ODS yet
> 
> ## Development
> If a commit is related to an issue (e.g. closes an issue) it should contain the issue number in the commit.
> ```

To add some context to the PR: 
We discussed about how to structure our workflow after our migration from GitLab and decided that instead of just deciding something, we want to make our decisions explicit and make them visible in the repository. This is the first draft and it is meant to be discussed here so we can improve it. 
Probably you have a point about the contributing an development section and we should stick to the conventions here. Issues like project boards, however, that do not really fit in one of these sections, so I am not sure on where to document them, or if we should leave them out altogether.

File: doc/github-workflow.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+## Repositories
+- Master branch is protected: No direct push, only via reviewed pull request from separate branches
+- If possible, pull requests should be linked to issues
+- Students work on their own forks using regular pull requests to get fast feedback
+
+## Branch Naming Scheme
+- *linked issue #*-*ods-subproject*-*short branch description*
+- example: 39-documentation-githubguidleines
+
+## Teams
```
[the remark this answers is not in the document]
Comment (mathiaszinnen) [reply doc/github-workflow.md:29]: Are you referring to the Teams section or to the naming scheme? 
I added both of them to make our choices explicit and enable discussion about them

File: doc/github-workflow.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+- Self staged issues may be discussed with the admin team to ensure later integration
+- For integration into ods codebase they need to sign a contributers agreement (via the CLA Assistant GitHub Plugin)
+- Code integration and review is to be done by pull requests. The pull requests have to be approved by at least one member of the repositories admins team. 
+
+## Repositories
+- Master branch is protected: No direct push, only via reviewed pull request from separate branches
+- If possible, pull requests should be linked to issues
+- Students work on their own forks using regular pull requests to get fast feedback
+
+## Branch Naming Scheme
```
[the remark this answers is not in the document]
Comment (georg-schwarz) [reply doc/github-workflow.md:25]: Agree

File: doc/github-workflow.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+- External contributions are to be encouraged
+- External contributors may stage issues or work on existing ones
+- Self staged issues may be discussed with the admin team to ensure later integration
+- For integration into ods codebase they need to sign a contributers agreement (via the CLA Assistant GitHub Plugin)
+- Code integration and review is to be done by pull requests. The pull requests have to be approved by at least one member of the repositories admins team. 
+
+## Repositories
+- Master branch is protected: No direct push, only via reviewed pull request from separate branches
+- If possible, pull requests should be linked to issues
+- Students work on their own forks using regular pull requests to get fast feedback
```
Comment (georg-schwarz) [inline doc/github-workflow.md:23]: I think we should not separate students from externals?

File: doc/github-workflow.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (mathiaszinnen) [reply doc/github-workflow.md:25]: Assuming you are talking about the branch naming: 
The naming scheme is reflecting what we are currently doing anyway. I personally think having the connected issue number and microservice makes it much easier to understand the context of the whole branch. You can get a lot of context information on the first look. 
Do you prefer to leave the branch naming up to the developer?

File: doc/github-workflow.md (hunk 5 above)
Comment (mathiaszinnen) [reply doc/github-workflow.md:23]: Yes, this applies to both externals and students. Maybe it is better to set common rules for both of them.

**Comment by mathiaszinnen** [comment]
Closing the PR since it is probably more productive to discuss this in person.

**Pull request closed without merging**